# Step 3. Evaluate Data Quality Issues in the Data Provided

### 1. read the data in dataframes

In [68]:
import pandas as pd

brands = pd.read_json('data/brands.json',lines=True)
receipts = pd.read_json('data/receipts.json',lines=True)
users = pd.read_json('data/users.json',lines=True)

###  2. take a look at the data 

In [4]:
brands.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

In [5]:
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [6]:
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

### 3. Check data quality

first, flatten the columns in brands table that has nested hash tables such as {'$oid': '5ff1e194b6a9d73a3a9f1052'}

In [69]:
brands['_id'] = brands['_id'].apply(lambda x: x['$oid'])
brands['cpg_ref'] = brands['cpg'].apply(lambda x: x['$ref'])
brands['cpg'] = brands['cpg'].apply(lambda x: x['$id']['$oid'])
brands.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                        cpg                       name  topBrand  \
0  601ac114be37ce2ead437550  test brand @1612366101024       0.0   
1  5332f5fbe4b03c9a25efd0ba                  Starbucks       0.0   
2  601ac142be37ce2ead437559  test brand @1612366146176       0.0   
3  601ac142be37ce2ead437559  test brand @1612366146051       0.0   
4  5332fa12e4b03c9a25efd1e7  test brand @1612366146827       0.0   

                       brandCode cpg_ref  
0                            NaN    Cogs  
1                      STARBUCKS    Cogs  
2  TEST BRANDCODE @1612366146176    Cogs  
3  TEST BRANDCODE @1612366146051    Cogs  
4  TEST BRANDCODE @1612366146827    Cogs

check basic information

In [70]:
brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           1167 non-null   object 
 1   barcode       1167 non-null   int64  
 2   category      1012 non-null   object 
 3   categoryCode  517 non-null    object 
 4   cpg           1167 non-null   object 
 5   name          1167 non-null   object 
 6   topBrand      555 non-null    float64
 7   brandCode     933 non-null    object 
 8   cpg_ref       1167 non-null   object 
dtypes: float64(1), int64(1), object(7)
memory usage: 82.2+ KB


In [71]:
brands['category'].value_counts()

Baking                         369
Beer Wine Spirits               90
Snacks                          75
Candy & Sweets                  71
Beverages                       63
Health & Wellness               44
Magazines                       44
Breakfast & Cereal              40
Grocery                         39
Dairy                           33
Condiments & Sauces             27
Frozen                          24
Personal Care                   20
Baby                            18
Canned Goods & Soups            12
Beauty                           9
Cleaning & Home Improvement      6
Deli                             6
Beauty & Personal Care           6
Bread & Bakery                   5
Dairy & Refrigerated             5
Household                        5
Outdoor                          1
Name: category, dtype: int64

In [48]:
brands['topBrand'].value_counts()

0.0    524
1.0     31
Name: topBrand, dtype: int64

In [47]:
brands['cpg_ref'].value_counts()

Cogs    1020
Cpgs     147
Name: cpg_ref, dtype: int64

check duplicates

In [72]:
# check if primary key is unique
brands[brands.duplicated(['_id'], keep=False)]

Empty DataFrame
Columns: [_id, barcode, category, categoryCode, cpg, name, topBrand, brandCode, cpg_ref]
Index: []

In [73]:
brands[brands.duplicated()]

Empty DataFrame
Columns: [_id, barcode, category, categoryCode, cpg, name, topBrand, brandCode, cpg_ref]
Index: []

check missing value

In [74]:
def check_missing(df):
    total_num = len(df)
    for column in df.columns:
        missing_count = df[column].isna().sum()
        missing_percent = round(missing_count/total_num,2)
        print(column, '\n missing_count: ', missing_count,'missing_percent: ',missing_percent)

In [75]:
check_missing(brands)

_id 
 missing_count:  0 missing_percent:  0.0
barcode 
 missing_count:  0 missing_percent:  0.0
category 
 missing_count:  155 missing_percent:  0.13
categoryCode 
 missing_count:  650 missing_percent:  0.56
cpg 
 missing_count:  0 missing_percent:  0.0
name 
 missing_count:  0 missing_percent:  0.0
topBrand 
 missing_count:  612 missing_percent:  0.52
brandCode 
 missing_count:  234 missing_percent:  0.2
cpg_ref 
 missing_count:  0 missing_percent:  0.0


So the data in 'brands' dataset are all unique, but there are some missing values. Some missing features inculd 'category', 'categoryCode',  'topBrand' and 'brandCode'. The 'categoryCode' has the most missing values, more than half of the record does not include 'categoryCode'. 

In [76]:
# flatten the user id which potentially can be a primary key
users['_id'] = users['_id'].apply(lambda x: x['$oid'])
users.head()

_id  active               createdDate  \
0  5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
1  5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
2  5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
3  5ff1e1eacfcf6c399c274ae6    True  {'$date': 1609687530554}   
4  5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

check basic information

In [78]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [79]:
users['active'].value_counts()

True     494
False      1
Name: active, dtype: int64

The user data are pretty imbalanced

In [27]:
users['role'].value_counts()

consumer       413
fetch-staff     82
Name: role, dtype: int64

In [80]:
users['signUpSource'].value_counts()

Email     443
Google      4
Name: signUpSource, dtype: int64

In [81]:
users['state'].value_counts()

WI    396
NH     20
AL     12
OH      5
IL      3
CO      1
KY      1
SC      1
Name: state, dtype: int64

check duplicates

In [77]:
users[users.duplicated(['_id'], keep=False)]

_id  active               createdDate  \
0    5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
1    5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
2    5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
3    5ff1e1eacfcf6c399c274ae6    True  {'$date': 1609687530554}   
4    5ff1e194b6a9d73a3a9f1052    True  {'$date': 1609687444800}   
..                        ...     ...                       ...   
490  54943462e4b07e684157a532    True  {'$date': 1418998882381}   
491  54943462e4b07e684157a532    True  {'$date': 1418998882381}   
492  54943462e4b07e684157a532    True  {'$date': 1418998882381}   
493  54943462e4b07e684157a532    True  {'$date': 1418998882381}   
494  54943462e4b07e684157a532    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
0    {'$date': 1609687537858}     consumer        Email    WI  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
3    {'$date': 1609687530597}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[353 rows x 7 columns]

So users dataset not only has duplicated user ids but also duplicated rows

In [84]:
print('duplicated user id rate:', round(len(users[users.duplicated(['_id'], keep=False)])/len(users),2))

duplicated user id rate: 0.71


check missing value

In [82]:
check_missing(users)

_id 
 missing_count:  0 missing_percent:  0.0
active 
 missing_count:  0 missing_percent:  0.0
createdDate 
 missing_count:  0 missing_percent:  0.0
lastLogin 
 missing_count:  62 missing_percent:  0.13
role 
 missing_count:  0 missing_percent:  0.0
signUpSource 
 missing_count:  48 missing_percent:  0.1
state 
 missing_count:  56 missing_percent:  0.11


The users dataset also has missing value in some features, inculding 'lastLogin', 'signUpSource', and 'state'. 

So far I have check the brands and users datasets and they both have data quality issue of missing values, and users dataset also contains about 71% of duplicated user Ids. 